contracts/external/cw-payroll/src/contract.rs

```diff
@@ -0,0 +1,598 @@
+        balance,
+        claimed_balance: Uint128::zero(),
+        start_time,
+        end_time,
+        rate_per_second,
+    };
+    let id = save_stream(deps, &stream)?;
+
+    let mut response = Response::new()
+        .add_attribute("method", "create_stream")
+        .add_attribute("stream_id", id.to_string())
+        .add_attribute("admin", admin.clone())
+        .add_attribute("recipient", recipient)
+        .add_attribute("balance", balance.to_string())
+        .add_attribute("start_time", start_time.to_string())
+        .add_attribute("end_time", end_time.to_string());
+
+    if refund > 0 {
+        // TODO: Change to support native and fix cw20
+        let cw20 = Cw20Contract(config.cw20_addr);
+        let msg = cw20.call(Cw20ExecuteMsg::Transfer {
+            recipient: owner.into(),
+            amount: refund.into(),
+        })?;
+
+        response = response.add_message(msg);
+    }
+    Ok(response)
+}
+
+// TODO: Add native balance recieve for diff streams
+pub fn execute_receive(
+    env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    wrapped: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    // TODO: Change based on asset type found
+    if config.cw20_addr != info.sender {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let msg: ReceiveMsg = from_binary(&wrapped.msg)?;
+    match msg {
+        ReceiveMsg::CreateStream {
+            admin,
+            start_time,
+            end_time,
+            recipient,
+        } => execute_create_stream(
+            env,
+            deps,
+            config,
+            // TODO: Change this, but still conform to standard
+            StreamParams {
+                admin: wrapped.sender,
+                recipient,
+                balance: wrapped.amount,
+                start_time,
+                end_time,
+                title: None,
+                description: None
+            },
+        ),
+    }
+}
+
+pub fn execute_distribute(
+    env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    id: u64,
+) -> Result<Response, ContractError> {
+    let mut stream = STREAMS
+        .may_load(deps.storage, id)?
+        .ok_or(ContractError::StreamNotFound {})?;
+
+    // TODO: Why is this required? is there a security issue here if someone pays to send you tokens all the time?
+    // NOTE: commenting out for simple testing with croncat, if required will add approved list.
+    // if stream.recipient != info.sender {
+    //     return Err(ContractError::NotStreamRecipient {
+    //         recipient: stream.recipient,
+    //     });
+    // }
+
+    // TODO: Def change to better comparitor
+    if stream.claimed_balance >= stream.balance {
+        return Err(ContractError::StreamFullyClaimed {});
+    }
+
+    let block_time = env.block.time.seconds();
+    let time_passed = std::cmp::min(block_time, stream.end_time).saturating_sub(stream.start_time);
```

Comment:
assuming this stream is validated correctly, we shouldn't need the `saturating_sub` i think? another reason i like checked types for this sort of stuff.